Add gender field to the personal information form in user account

aramboi opened the pull request:
Closes #1200
Closes #1210

aramboi committed:
Add gender field to the personal information form in user account

Closes #1200

umgelurgel commented on conference/models.py:
```diff
@@ -336,9 +343,15 @@ class AttendeeProfile(models.Model):
+        ),
+    )
+    gender = models.CharField(
+        max_length=1, choices=ATTENDEEPROFILE_GENDER, blank=True, default=""
```
Regarding the default, I'm wondering whether it's worth it to distinguish between users who have yet to provide a value and users who selected `prefer not to say`. WDYT?

aramboi replied:
I made it so we have also blank as default.

umgelurgel commented:
Looks good so far! Could you also attach a screenshot of the form after adding the styling changes?

europython-society commented:
There definitely should be a distinction between "we don't know"
and "prefer not to say".

On 18.02.2020 09:38, umgelurgel wrote:
> *@umgelurgel* commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In conference/models.py
> <https://github.com/EuroPython/epcon/pull/1201#discussion_r380524100>:
> 
>> @@ -336,9 +343,15 @@ class AttendeeProfile(models.Model):
>      phone = models.CharField(
>          _('Phone'),
>          max_length=30, blank=True,
> -        help_text=_('Enter a phone number where we can contact you in case of administrative issues.<br />Use the international format, eg: +39-055-123456'),
> +        help_text=_(
> +            'Enter a phone number where we can contact you in case of '
> +            'administrative issues.<br>Use the international format, '
> +            'eg: +39-055-123456'
> +        ),
> +    )
> +    gender = models.CharField(
> +        max_length=1, choices=ATTENDEEPROFILE_GENDER, blank=True, default=""
> 
> Regarding the default, I'm wondering whether it's worth it to
> distinguish between users who have yet to provide a value and users who
> selected |prefer not to say|. WDYT?
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/EuroPython/epcon/pull/1201?email_source=notifications&email_token=ACOLSRMEW3LNFEUCHNP2IMLRDONBZA5CNFSM4KWZ7OEKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCV3XHDI#pullrequestreview-360149901>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACOLSROKUOXWBDUGBCYNW6DRDONBZANCNFSM4KWZ7OEA>.
> 

-- 
Marc-Andre Lemburg
EuroPython Society Chair
http://www.europython-society.org/
http://www.malemburg.com/

artcz commented:
I belive the same model is used as backend for the CFP form, where last year we had a free text form instead of a dropdown.

https://github.com/EuroPython/epcon/blob/c171753ea4738dc8fcc5fe023427d1ce19280e35/conference/cfp.py#L309-L313

That also means that there are more values in that field already
```
> select count(distinct gender) from conference_attendeeprofile;
19
```

malemburg commented:
I checked the database: the existing gender field would probably serve as better template for default values than the t-shirt size. Because the system (used to) default to men's L for t-shirt sizes, we have quite a few entries which don't match the name or gender setting in the profile.

malemburg commented:
We'd have to change the existing gender field to a drop-down, though, also in the CFP form and normalize the database.

malemburg commented:
Hmm, looking at the numbers, we still need some form of logic which requires users to set the gender when they log in to the site. We only have the information from speakers and not from regular attendees.

aramboi committed:
Update CFP page to include gender dropdown.

Allow the gender field to be blank.
Add migration file to use choices for gender.

aramboi commented:
Changed in the CFP page as well and added a blank default in case they choose not to answer at all. Below are some screenshots with the pages.

![Screenshot_2020-02-19 User's Panel EuroPython 2020 · Dublin, Ireland, 20-26 July 2020](https://user-images.githubusercontent.com/12439/74877428-b31c4c80-5365-11ea-8056-7734e47c1c32.png)
![Screenshot_2020-02-19 EuroPython 2020 · Dublin, Ireland, 20-26 July 2020](https://user-images.githubusercontent.com/12439/74877444-ba435a80-5365-11ea-848b-4221c898196d.png)

aramboi commented:
@artcz @umgelurgel can you take a look and let me know if I missed any other place where gender should be present in a form or template? Otherwise I would merge this so we can have it for launch.

umgelurgel reviewed (approved):
LGTM to merge! The data on the DB will still need to be cleaned up prior to deployment, I think.

umgelurgel commented:
@aramboi I'm leaving this umerged for now - I think it's best to merge it only once the gender data in staging is normalised, so that this can be deployed.

malemburg commented:
On 27.02.2020 19:26, umgelurgel wrote:
> @aramboi <https://github.com/aramboi> I'm leaving this umerged for now -
> I think it's best to merge it only once the gender data in staging is
> normalised, so that this can be deployed.

Hi Angel,

it's better to start from scratch. The existing data is unreliable.

Thanks,
-- 
Marc-Andre Lemburg
http://www.malemburg.com/

malemburg commented on conference/user_panel.py:
```diff
@@ -285,13 +286,23 @@ class ProfileSettingsForm(forms.ModelForm):
+        ),
+        choices=(("", "---", "---"),) + ATTENDEEPROFILE_GENDER,
+        widget=forms.Select,
+        required=False,
```
This field needs to be "required".

aramboi replied:
Will do.

malemburg commented on conference/cfp.py:
```diff
@@ -300,15 +301,20 @@ class AddSpeakerToTalkForm(forms.ModelForm):
-    gender = forms.CharField(
-        label="Type your gender",
-        max_length=32,
+    gender = forms.ChoiceField(
```
Do we really need this field now that we have a user field for gender ? People submitting talks will have to have a user account.

aramboi replied:
Right. I will remove it entirely.

malemburg commented on tests/test_cfp.py:
```diff
@@ -189,7 +189,7 @@ def test_if_user_can_add_a_speaker_to_a_proposal(user_client):
             "users_family_name": "Doe",
             "phone": "+48523456789",
-            "gender": "M",
+            "gender": "m",
```
See above. Not needed.

malemburg replied:
Here's the ticket for getting the users to fill in the new field: https://github.com/EuroPython/epcon/issues/1210

aramboi replied:
Got it.

malemburg reviewed (changes requested):
Hi Angel, thanks for the PR, but I think it still needs a few updates (see review comments). I'll open a new ticket for making sure that users update their profile in case the gender field is not set.

matrixise commented on conference/models.py:
```diff
@@ -309,9 +309,16 @@ def getOrCreateForUser(self, user):
-    ('p', "PUBLIC", 'Publicly available'),
-    ('m', "PARTICIPANTS_ONLY", 'Visible to EuroPython attendees'),
-    ('x', "PRIVATE", 'Visible only to you'),
+    ("p", "PUBLIC", "Publicly available"),
```
This change is not useful in this case. Just add the needed code

aramboi replied:
This code block had inconsistent use of quotes, I just made them all the same.

matrixise replied:
@aramboi sure, but this kind of modification could be done in another commit. just to avoid to review the useless modifications to the bug fix

aramboi replied:
I agree with you and this is the way we should do it given enough resources. At this moment I would rather get this in than create more overhead for the people involved.

aramboi committed:
Merge branch 'dev/ep2020' into ar/issue-1200

aramboi committed:
Enforce profile update if required field is missing

aramboi committed:
Merge branch 'dev/ep2020' into ar/issue-1200

aramboi committed:
Add redirect test in case gender is missing

aramboi committed:
Add profile check on cart and cfp paths. Add more tests.

aramboi commented:
@malemburg @umgelurgel @matrixise Made the changes required in both tickets. Please review.

Right now we check if gender is present on the cfp, cart and user dashboard pages. If it is not, we redirect the user to the user settings and post a message to complete their profile.

malemburg commented:
Looks good. Thanks, Angel.
